jumanji/environments/logic/sudoku/utils.py

```diff
@@ -0,0 +1,99 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import chex
+import jax
+import jax.numpy as jnp
+
+from jumanji.environments.logic.sudoku.constants import BOARD_WIDTH, BOX_IDX
+
+
+def apply_action(action: chex.Array, board: chex.Array) -> chex.Array:
+    return board.at[action[0], action[1]].set(action[2])
+
+
+def puzzle_completed(board: chex.Array) -> chex.Array:
+    """Checks that every row, column and 3x3 boxes includes all figures.
+
+    Args:
+        board: The sudoku board.
+
+    Returns:
+        condition: A `bool` indicator that validates if the puzzle is solved.
+    """
+
+    def _validate_row(row: chex.Array) -> chex.Array:
+        condition = jnp.sort(row) == jnp.arange(BOARD_WIDTH)
+        return condition.all()
+
+    condition_rows = jax.vmap(_validate_row)(board).all()
+    condition_columns = jax.vmap(_validate_row)(board.T).all()
+    condition_boxes = jax.vmap(_validate_row)(jnp.take(board, jnp.array(BOX_IDX))).all()
+    return condition_rows & condition_columns & condition_boxes
+
+
+def validate_board(board: chex.Array) -> chex.Array:
+    """Checks that every row, column and 3x3 boxes contain no duplicate number.
+
+    Args:
+        board: The sudoku board.
+
+    Returns:
+        condition: A `bool` indicator that validates a board or not.
+    """
+
+    def _validate_row(row: chex.Array) -> chex.Array:
+        return jax.nn.one_hot(row, BOARD_WIDTH).sum(axis=0).max() <= 1
+
+    condition_rows = jax.vmap(_validate_row)(board).all()
+    condition_columns = jax.vmap(_validate_row)(board.T).all()
+    condition_boxes = jax.vmap(_validate_row)(jnp.take(board, jnp.array(BOX_IDX))).all()
+
+    return condition_rows & condition_columns & condition_boxes
+
+
+def update_action_mask(action_mask: chex.Array, board: chex.Array) -> chex.Array:
+    """Updates the action mask according to the current board state.
+
+    Args:
+        action_mask: The action mask.
+        board: The sudoku board.
+
+    Returns:
+        The updated action mask.
+    """
+    row_mask = 1 - jax.nn.one_hot(board, BOARD_WIDTH).any(axis=1) * 1
+    column_mask = 1 - jax.nn.one_hot(board.T, BOARD_WIDTH).any(axis=1) * 1
+
+    boxes = board.reshape(BOARD_WIDTH**2).take(jnp.array(BOX_IDX))
+    box_mask = 1 - jax.nn.one_hot(boxes, BOARD_WIDTH).any(axis=1) * 1
+
+    boxes_action_mask = action_mask.reshape(BOARD_WIDTH**2, BOARD_WIDTH)[
+        jnp.array(BOX_IDX)
+    ]
+    boxes_action_mask *= box_mask.reshape(BOARD_WIDTH, 1, BOARD_WIDTH)
```

Comment:
I am probably missing something, but why do we have to multiply it by the old mask ? Isn't `boxes_action_mask = box_mask.reshape(BOARD_WIDTH, 1, BOARD_WIDTH)` enough ?

Reply:
`box_mask.reshape(BOARD_WIDTH, 1, BOARD_WIDTH)` indicates for each box the remaining digits writable (but has no prior among which cell empty) and `action_mask.reshape(BOARD_WIDTH**2, BOARD_WIDTH)[BOX_IDX]` indicates which cells are empty. By multiplying both you obtain a mask that tells you which cells are empty and for each of these cells which digits are writable (w.r.t the box constraint).

Reply:
Yes good :+1: